### PR TITLE
Improve BigDecimalGenerator to cover both positive and negative range

### PR DESCRIPTION
**Pull request by deepakhalale**

### CODE REVIEW

File: test-utils/test-utils-core/src/main/java/uk/gov/justice/services/test/utils/core/helper/TypeCheck.java [hunk 1]
```diff
@@ -1,18 +1,20 @@
...
 
-import uk.gov.justice.services.test.utils.core.random.Generator;
-
-import java.util.function.Function;
-
 import static java.util.Objects.requireNonNull;
 import static org.hamcrest.MatcherAssert.assertThat;
 import static org.hamcrest.core.Is.is;
 
+import uk.gov.justice.services.test.utils.core.random.Generator;
```
Comment (matt-rich) [inline test-utils/test-utils-core/src/main/java/uk/gov/justice/services/test/utils/core/helper/TypeCheck.java]: Do we still have issues with imports being reordered, is everyone using the same style guide?

File: test-utils/test-utils-core/src/main/java/uk/gov/justice/services/test/utils/core/helper/TypeCheck.java [hunk 2]
```diff
@@ -46,7 +48,7 @@ public void verify(Times times) {
             if (preCondition != null && !preCondition.apply(next)) {
                 i--;
             } else {
-                assertThat(condition.apply(next), is(true));
+                assertThat(String.format("failed on attempt %s for value %s", i + 1, next), condition.apply(next), is(true));
```
Comment (matt-rich) [inline test-utils/test-utils-core/src/main/java/uk/gov/justice/services/test/utils/core/helper/TypeCheck.java]: static import format

File: test-utils/test-utils-core/src/main/java/uk/gov/justice/services/test/utils/core/random/BigDecimalGenerator.java [hunk 3]
```diff
@@ -1,23 +1,56 @@
...
-        final double pow = pow(10, decimalPlaces);
-        return new BigDecimal((long) RANDOM.nextInt((int) (max * pow)) / pow).setScale(decimalPlaces, FLOOR);
+        return generateRandomBigDecimalBetween(min, max).setScale(scale, ROUND_HALF_EVEN);
+    }
+
+    private BigDecimal generateRandomBigDecimalBetween(final BigDecimal min, final BigDecimal max) {
+        if (min.compareTo(max) >= 0) {
+            throw new IllegalArgumentException(format("Min value cannot be greater than or equal to Max value, got Min: %s and Max: %s", min, max));
+        }
+        if (scale < 0) {
```
Comment (matt-rich) [inline test-utils/test-utils-core/src/main/java/uk/gov/justice/services/test/utils/core/random/BigDecimalGenerator.java]: This isn't really set in this method, should it blow up on construction?

File: test-utils/test-utils-core/src/main/java/uk/gov/justice/services/test/utils/core/helper/TypeCheck.java (hunk 1 above)
Comment (deepakhalale) [inline test-utils/test-utils-core/src/main/java/uk/gov/justice/services/test/utils/core/helper/TypeCheck.java]: Looks like imports weren't organised earlier!

File: test-utils/test-utils-core/src/main/java/uk/gov/justice/services/test/utils/core/random/BigDecimalGenerator.java (hunk 3 above)
Comment (deepakhalale) [inline test-utils/test-utils-core/src/main/java/uk/gov/justice/services/test/utils/core/random/BigDecimalGenerator.java]: Moved to constructor to fail fast

File: test-utils/test-utils-core/src/main/java/uk/gov/justice/services/test/utils/core/helper/TypeCheck.java (hunk 2 above)
Comment (deepakhalale) [inline test-utils/test-utils-core/src/main/java/uk/gov/justice/services/test/utils/core/helper/TypeCheck.java]: Done.

File: test-utils/test-utils-core/src/test/java/uk/gov/justice/services/test/utils/core/random/RandomGeneratorTest.java [hunk 4]
```diff
@@ -81,11 +91,12 @@ public void shouldGenerateRandomPercentage() {
...
         final Generator<BigDecimal> percentageGenerator = RandomGenerator.PERCENTAGE;
 
-        // when
-        typeCheck(percentageGenerator, s -> percentageGenerator.next().compareTo(percentageGenerator.next()) != 0).verify(times(NUMBER_OF_TIMES));
+        // when & then
+        typeCheck(percentageGenerator, s -> s.toPlainString().matches(PERCENTAGE_PATTERN)).verify(times(NUMBER_OF_TIMES));
     }
 
     @Test
+    @Ignore("Broken test. Need to fix.")
```
Comment (allanmckenzie) [inline test-utils/test-utils-core/src/test/java/uk/gov/justice/services/test/utils/core/random/RandomGeneratorTest.java]: Should this be ignored? Maybe better to either fix of delete

Comment (deepakhalale) [inline test-utils/test-utils-core/src/test/java/uk/gov/justice/services/test/utils/core/random/RandomGeneratorTest.java]: This will be fixed as part of different PR which I am working at the moment.
